examples/01-fails.yaml

```diff
@@ -65,3 +65,20 @@ tests:
       - exec: node1
         run: /bin/true
         expect: throw("Intentional Failure")
+
+  fail-with-outputs:
+    name: Failing test that defines outputs
+    outputs:
+      FAILED_TEST_OUTPUT: ${{ 1 + 1 }}
+    steps:
+      - exec: node1
+        run: /bin/false
+
+  maybe-check-empty-failure-outputs:
+    depends: fail-with-outputs
+    name: Assert that outputs should be empty from failed tests
+    steps:
+      - exec: node1
+        run: /bin/true
+        expect:
+          - tests['fail-with-outputs'].outputs == {}
```

Comment:
What if we have output from a fail and want to use that later (--continue-on-error and `if: failure()`)?

Also, would this also be "{}" if the user had an empty output map but it succeeded? Maybe it should be nil (i.e. something falsey) on failure?

Reply:
Good points.  I didn't expect running outputs after a failed test (or step) to be useful, as something "upstream" must have gone wrong (`env` interpolation threw an error, `run` didn't provide the expected result, etc), but maybe I'm wrong about that.  Perhaps we should still attempt to interpolate the outputs regardless, and if there's another error, so be it?  (Btw, `--continue-on-error` would relate to tests observing other [failed] tests' outputs, whereas right now we only support `if: failure()` on steps, but I get your point).

Quick related question to probe how you'd expect things to work.  Right now, if `env` interpolation throws an error, we fail the test or step and do not run the steps or `run` command respectively.  If we switch things such that `outputs` are always interpolated (except if `if` is falsy), do you still think we have the correct behavior in that `env`-stopping-`run` case?  Do you think we should swallow `env` interpolation errors, returning nil or an empty string or something, then proceed with the rest of the test/step execution?  To be clear, I think our current behavior is the right thing (albeit tricky to implement, it seems), and I'm also ok with `outputs` being a sort of exception that always runs, especially since the user would have some control here, ex: `outputs: { FOO: '${{ (failure() && ..) || ... }}' }`.

I absolutely agree with a failed test returning nil instead of "{}" - thank you, I'll fix that now.  And you're right that a successful test will return "{}" regardless of whether it defined an outputs map or not.  Perhaps I should add a test for that as well.

Reply:
Ok updated to dissoc outputs in the failing case and added a successful test outputs test.  Let me know what you think about running outputs on failed tests (which will probably reverse the decision to return nil/falsy outputs on failed tests).

Reply:
Maybe I'm misinterpreting the test. It looks to me like `maybe-check-empty-failure-outputs` test is not getting output because `fail-with-outputs` calls `false` which fails the step (and therefore the test). So my concern is that if instead of `false`, a program is called that returns some stderr output and then fails. It seems like a later test that depends on that should be able to access/check that output somehow. An error during the actual interpolation of env/outputs should return null (but I'm not sure I see that case being tested).